pygments/lexers/q.py

```diff
@@ -0,0 +1,228 @@
+            (r"'`[a-zA-Z]*(\w|[.])*", Name.Exception),
+            # File Symbols
+            (r"`:[a-zA-Z/]*(\w|[./])*", String.Symbol),
+            # Symbols
+            (r"[`][a-zA-Z]*(\w|[.])*", String.Symbol),
+            # Numbers
+            include("numbers"),
+            # Variable Names
+            (r"[a-zA-Z](\w|[.])*", Name),
+            # Operators
+            (r"[-=+*#$%@!~^&:.,<>'\\|/?_]", Operator),
+            # Punctuation
+            (r"[;]", Punctuation),
+        ],
+        "functions": [
+            include("root"),
+            (r"\}", Punctuation, "#pop"),
+        ],
+        "parentheses": [
+            include("root"),
+            (r"\)", Punctuation, "#pop"),
+        ],
+        "brackets": [
+            include("root"),
+            (r"\]", Punctuation, "#pop"),
+        ],
+        "numbers": [
+            (r"[01]+b", Number.Bin),
+            (r"0[nNwW][cefghijmndzuvtp]?", Number),  # nulls/infinities
+            (
+                (
+                    r"(?:[0-9]{4}[.][0-9]{2}[.][0-9]{2}|[0-9]+)"
+                    "D(?:[0-9](?:[0-9](?:[:][0-9]{2}"
+                    "(?:[:][0-9]{2}(?:[.][0-9]*)?)?)?)?)?"
+                ),
+                Literal.Date,  # timestamp
+            ),
+            (
+                (
+                    r"[0-9]{4}[.][0-9]{2}"
+                    "(?:m|[.][0-9]{2}(?:T(?:[0-9]{2}:[0-9]{2}"
+                    "(?:[:][0-9]{2}(?:[.][0-9]*)?)?)?)?)"
+                ),
+                Literal.Date,  # datetime
+            ),
+            (
+                (r"[0-9]{2}:[0-9]{2}" "(?:[:][0-9]{2}" "(?:[.][0-9]{1,3})?)?"),
+                Literal.Date,  # time
+            ),
+            (
+                r"[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}",
+                Number.Hex,  # GUID
+            ),
+            (r"0x[0-9a-fA-F]+", Number.Hex),
+            (r"[0-9]+[jnp]", Number.Integer.Long),
+            (r"[0-9]+[ihtuv]", Number.Integer),
+            (r"([0-9]*[.]?[0-9]+|[0-9]+[.]?[0-9]*)[eE][+-]?[0-9]+[ef]?", Number.Float),
+            (r"([0-9]*[.][0-9]+|[0-9]+[.][0-9]*)[ef]?", Number.Float),
+            (r"[0-9]+[ef]", Number.Float),
+            (r"[0-9]+[c]", Number),
+            (r"[0-9]+", Number.Integer.Long),
+        ],
+        "comments": [
+            (r"[^\\]", Comment.Multiline),
+            (r"^\\", Comment.Multiline, "#pop"),
+            (r"[\\]", Comment.Multiline),
+        ],
+        "strings": [
+            (r"\\.", String.Double),
+            (r"[^\"]", String.Double),
+            (r"\"", String.Double, "#pop"),
+        ],
+    }
+
+
+class QLexer(KLexer):
+    """
+    For `Q <http://kx.com/>`_ source code.
+
+    .. versionadded:: 2.12
+    """
+
+    name = "Q"
+    aliases = ["q"]
+    filenames = ["*.q"]
+
+    tokens = {
+        "root": [
+            include("whitespace"),
+            include("keywords"),
+            include("macros"),
+            include("declarations"),
+        ],
```

Comment:
Maybe you can simplify that as

```
"root": [
  (words(...), Operator.Word),
  inherit,
]
```

?

Reply:
if there are keywords in comments (or strings) , they would be tokenized first, and then the comments (or strings) would not highlight properly.  i'll add a test case to ensure this is caught.